crates/sui-core/src/authority_server.rs

```diff
@@ -519,6 +524,97 @@ impl ValidatorService {
             .into());
         }
 
+        let request = request.into_inner();
+        let transaction = bcs::from_bytes::<Transaction>(&request.transaction).map_err(|e| {
+            SuiError::TransactionDeserializationError {
+                error: e.to_string(),
+            }
+        })?;
+        transaction.validity_check(epoch_store.protocol_config(), epoch_store.epoch())?;
+
+        // Check system overload
+        let overload_check_res = self.state.check_system_overload(
+            &*consensus_adapter,
+            transaction.data(),
+            state.check_system_overload_at_signing(),
+        );
+        if let Err(error) = overload_check_res {
+            metrics
+                .num_rejected_tx_during_overload
+                .with_label_values(&[error.as_ref()])
+                .inc();
+            return Err(error.into());
+        }
+
+        let _handle_tx_metrics_guard = metrics.handle_submit_transaction_latency.start_timer();
+
+        let transaction = {
+            let _metrics_guard = metrics.tx_verification_latency.start_timer();
+            epoch_store.verify_transaction(transaction).tap_err(|_| {
+                metrics.signature_errors.inc();
+            })?
+        };
+
+        // Enable Trace Propagation across spans/processes using tx_digest
+        let tx_digest = transaction.digest();
+        let _span = error_span!("validator_state_submit_transaction", ?tx_digest);
+
+        // TODO(fastpath): Skip checking for execution effects
+        // Effects are useless in mfp if they cannot be mapped to a consensus
+        // position
+        let _ = state
+            .handle_vote_transaction(&epoch_store, transaction.clone())
+            .tap_err(|e| {
+                if let SuiError::ValidatorHaltedAtEpochEnd = e {
+                    metrics.num_rejected_tx_in_epoch_boundary.inc();
+                }
+            })?;
+
+        let _latency_metric_guard = metrics
+            .handle_submit_transaction_consensus_latency
+            .start_timer();
+        let span = error_span!("submit_transaction", tx_digest = ?transaction.digest());
+        self.handle_submit_to_consensus_for_position(
+            nonempty![ConsensusTransaction::new_user_transaction_message(
+                &self.state.name,
+                transaction.into()
+            )],
+            &epoch_store,
+        )
+        .instrument(span)
+        .await
+        .and_then(|(mut resp, spam_weight)| {
+            // Only submitting a single tx so we should get back a single consensus position
+            let consensus_position = resp.remove(0);
+
+            let submit_transaction_response = RawSubmitTxResponse::into_raw(consensus_position)?;
+
+            Ok((
+                tonic::Response::new(submit_transaction_response),
+                spam_weight,
+            ))
+        })
+    }
+
+    async fn handle_get_effects(
```

Comment:
We probably do want to use wait_for_effects since it handles the wait within the validator, instead of client keeping polling and wait.
On a side note the function name `handle_vote_transaction` is a bit strange in AuthorityState. I seems to be literally getting executed effects.